keps/sig-node/4580-deprecate-kubelet-runonce/README.md

```diff
@@ -0,0 +1,355 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Deprecated and removed kubelet support for RunOnce mode.
+
+## Motivation
+
+* Runonce does not work in systemd mode.
+* Runonce mode doesn't support many newer pod features (init containers).
+* The pod lifecycle for runonce mode is even less well defined than normal.
+* The mode appears to be entirely untested in CI, except for a few unit tests.
+
+### Goals
+
+* Mark the `RunOnce` field in `KubeletConfiguration` and the `runonce` flag of kubelet as deprecated,  and finally remove the runonce flag.
+* Remove kubelet support for RunOnce mode.
+
+### Non-Goals
+
+Immediate removal: the deprecation and removal process will be gradual and feature gate to increase awareness among potential users.
+
+## Proposal
+
+The runonce mode of kubelet will exit the kubelet process after spawning pods from the local manifests or remote URL. It is suitable for scenarios where one-time tasks need to be run on the node, this proposal outlines plans to deprecate and remove runonce mode in kubelet.
+
+### Risks and Mitigations
+
+Some people may still rely on this feature, but podman addresses the same use case with more well-supported way, ref: https://docs.podman.io/en/latest/markdown/podman-kube.1.html. Affected users can migrate to *podman kube subcommand* on demand.
```

Comment:
What about users of docker or other runtimes?

Reply:
Do you have at least a rough estimate how many users this might affect?

Reply:
> Do you have at least a rough estimate how many users this might affect?

I can't estimate how many users this will affect, and in fact a Google search yields very few blogs or discussions about the Runonce model.



> What about users of docker or other runtimes?

Maybe they can install a podman, podman and docker do not conflict.

Reply:
Let's make sure we call out that affected users have alternatives mentioning both podman and docker.

Reply:
Because Docker does not support this mode and there is no reliable third-party tool available, I added a description of manually translating Kubernetes YAML to Docker Compose files. Does this meet our expectations?